### PR TITLE
Deploy dqt notifications resources

### PR DESCRIPTION
**Pull request by enformatique11**
**Context **

Tra-dqt-notifications deployment

**Changes proposed in this pull request ** 
Configured terraform code for the tra-dqt-notifications resources as follows:

•	App Service Plan
•	Function App
•	MSSQL Server
•	MSSQL Database
•	Servicebus Namespace
•	Servicebus Topic
•	Storage Account

**Guidance to review**
Can you please check the terraform for DFE-Digital/tra-dqt-notifications:
- Deploy manually: `make dev terraform-apply`

**Checklist**

•	_Tested in GitHub Action – the workflow – deployed in dev environment – Successfully
•	Rebased branch - deploy resources
•	Cleaned commit history_

### CODE REVIEW

**Comment by enformatique11** [comment]
The PR is linked to the previous PR https://github.com/DFE-Digital/tra-dqt-notifications/pull/25
which has been closed by mistake.

**Comment by enformatique11** [comment]
Link to resources deployed: https://portal.azure.com/?Microsoft_Azure_PIMCommon=true#@platform.education.gov.uk/resource/subscriptions/9816b7b0-58ca-4972-b25b-c7dbd3ee1c6d/resourceGroups/s165d01-dqtnoti-dv-rg/overview

**Comment by enformatique11** [comment]
The tags have been updated as per request- 
Product: Database of Qualified Teachers
Service Offering: Database of Qualified Teachers

**Comment by NickGraham101** [comment]
@enformatique11 looks like you've deleted most of the contents of `/DqtNotifications` folder, that needs to be restored.

File: .github/workflows/build-and-deploy-notifications.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
+name: Build and Deploy
+
+on:
+  push:
+    branches:
+    - main
+    - deploy-resources
```
Comment (NickGraham101) [inline .github/workflows/build-and-deploy-notifications.yml:7]: this trigger needs deleting, it was only added for testing purposes

File: .github/workflows/build-and-deploy-notifications.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  pull_request:
+    branches:
+    - main
+
+  workflow_dispatch:
+
+concurrency:
```
Comment (NickGraham101) [inline .github/workflows/build-and-deploy-notifications.yml:14]: can we move this concurrency group to the build job.  on other projects it's caused the terraform statefile to get locked

File: .github/workflows/build-and-deploy-notifications.yml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  build:
+    name: Build & test
```
Comment (NickGraham101) [inline .github/workflows/build-and-deploy-notifications.yml:20]: There's quite a lot missing from the build job, can you copy over everything from the version of build.yml that's currently in the `main` branch

File: .github/workflows/build-and-deploy-notifications.yml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    #if: github.ref == 'refs/heads/main' && github.event_name == 'push'
+    concurrency: deploy_${{ matrix.environment }}
+    needs: [build]
+    strategy:
+      max-parallel: 1
+      matrix:
+        environment: [dev] #[dev, test, preprod, production]
+    environment:
+      name: ${{ matrix.environment }}
+      #url: ${{ steps.deploy.outputs.environment_url }}
```
Comment (NickGraham101) [inline .github/workflows/build-and-deploy-notifications.yml:53]: these 3 lines need uncommenting

File: terraform/app.tf [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+# Servicebus Topic
+resource "azurerm_servicebus_topic" "servicebus_topic" {
+  name                = local.servicebus_topic_name
+  namespace_id        = azurerm_servicebus_namespace.servicebus_namespace.id
+  enable_partitioning = true
+}
+
+#Storage account
+resource "azurerm_storage_account" "st" {
```
Comment (NickGraham101) [inline terraform/app.tf:75]: Can we change the name to something more descriptive, `storage_account` is fine

File: terraform/app.tf [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+  }
+}
+
+#MSSQL Server
+resource "azurerm_mssql_server" "mssql_server" {
+  name                         = local.mssql_server
+  resource_group_name          = data.azurerm_resource_group.resource_group.name
+  location                     = data.azurerm_resource_group.resource_group.location
+  version                      = "12.0"
+  administrator_login          = local.administrator_login
```
Comment (NickGraham101) [inline terraform/app.tf:106]: Lets put the admin login in a secret as well, it'll be more secure

File: terraform/app.tf [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+  administrator_login_password = local.infrastructure_secrets.SQL_ADMIN_PASSWORD
+}
+
+#MSSQL Database
+resource "azurerm_mssql_database" "mssql_database" {
+  name        = local.mssql_database
+  server_id   = azurerm_mssql_server.mssql_server.id
+  collation   = "SQL_Latin1_General_CP1_CI_AS"
```
Comment (NickGraham101) [inline terraform/app.tf:114]: @gunndabad do we know what collation the production db is using?  This is the default for `English (United States)`

File: terraform/providers.tf [hunk 8]
```diff
@@ -0,0 +1,14 @@
+locals {
+  azure_credentials      = try(jsondecode(var.azure_sp_credentials_json), null)
+  infrastructure_secrets = yamldecode(data.azurerm_key_vault_secret.secrets["INFRASTRUCTURE"].value)
+  }
```
Comment (NickGraham101) [inline terraform/providers.tf:4]: Fix formatting, this curly bracket shouldn't be indented

File: terraform/variables.tf [hunk 9]
```diff
@@ -0,0 +1,133 @@
+variable "project" {
```
Comment (NickGraham101) [inline terraform/variables.tf:1]: Delete, this isn't being used

File: terraform/variables.tf [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+locals {
+  hosting_environment         = var.environment_name
+  servicebus_namespace_name   = "${var.namespace_prefix}dqtnoti-${var.environment_name}-sbn"
+  servicebus_topic_name       = "${var.topic_prefix}dqtnoti-${var.environment_name}-sbt"
+  storage_account             = "${var.storage_account}dqtnoti${var.environment_name}sg"
+  app_service_plan_name       = "${var.plan_prefix}dqtnoti-${var.environment_name}-spl"
+  azurerm_function_app_name   = "${var.function_app_name}dqtnoti-${var.environment_name}-fapp"
+  mssql_server                = "${var.mssql_server}dqtnoti-${var.environment_name}-mssql"
+  mssql_database              = "${var.mssql_database}dqtnoti-${var.environment_name}-sqldb"
+  administrator_login         = "${var.mssql_server}dqtnoti-${var.environment_name}-admin"
```
Comment (NickGraham101) [inline terraform/variables.tf:130]: remove this and store in KeyVault secret

File: terraform/variables.tf [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+  default = ""
+}
+
+variable "azure_sp_credentials_json" {
+  type    = string
+  default = null
+}
+
+variable "resource_group_name" {
+type        = string
```
Comment (NickGraham101) [inline terraform/variables.tf:59]: Fix indenting

File: terraform/variables.tf [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+  variable "storage_account" {
+  type    = string
+  default = ""
+}
+
+variable "key_vault_name" {
+  type = string
+}
+
+variable "plan_prefix" {
```
Comment (NickGraham101) [inline terraform/variables.tf:48]: Can this be more descriptive, eg `app_service_plan_prefix`

File: terraform/variables.tf [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+variable "azure_sp_credentials_json" {
+  type    = string
+  default = null
+}
+
+variable "resource_group_name" {
+type        = string
+}
+
+variable "sku" {
```
Comment (NickGraham101) [inline terraform/variables.tf:62]: Can this be more descriptive, should say what resource it's the sku for, eg `service_bus_sku`

File: terraform/variables.tf [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+  type        = string
+  default     = "Standard"
+  description = "SKU type (Basic, Standard and Premium)"
+}
+
+variable "capacity" {
```
Comment (NickGraham101) [inline terraform/variables.tf:68]: Can this be more descriptive, which resource is it the capacity of?

File: terraform/variables.tf [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+variable "function_app_name" {
+  type    = string
+  default = ""
+}
+
+variable "storage_account_name" {
+  type = string
+}
+
+variable "zone_redundant" {
```
Comment (NickGraham101) [inline terraform/variables.tf:83]: Needs to be prefixed with the resource it relates to

File: terraform/variables.tf [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+  type        = bool
+  default     = false
+  description = "Whether or not this resource is zone redundant (true or false)"
+}
+
+variable "enable_public_access" {
```
Comment (NickGraham101) [inline terraform/variables.tf:89]: Doesn't look like this is used, delete it it's not

File: terraform/variables.tf [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+  type        = bool
+  default     = false
+  description = "Enable public access (should only be enabled for a migration when using the Premium SKU and a private endpoint connection)"
+}
+
+variable "enable_private_endpoint" {
+  default     = false
+  description = "Enable Private endpoint? Only available with the Premium SKU, if set to true a Premium type Service Bus Namespace will be deployed automatically"
```
Comment (NickGraham101) [inline terraform/variables.tf:97]: [Looks](https://github.com/DFE-Digital/tra-dqt-notifications/blob/dd2086d87b5de6d66fc39b21ed576d1111823aca/terraform/app.tf#L3) like a Standard sku is deployed if this is set to true

Is enabling private endpoint a requirement?

File: terraform/variables.tf [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+  type    = number
+}
+variable "mssql_max_size_gb" {
+  description = "The max size of the database in gigabytes"
+  type        = number
+  default     = 2
+}
+
+variable "mssql_database" {
+  description = "The name of the MSSQL database to create. Must be set if `enable_mssql_database` is true"
```
Comment (NickGraham101) [inline terraform/variables.tf:111]: `enable_mssql_database` doesn't exist

File: terraform/variables.tf [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+  type        = string
+  default     = ""
```
Comment (NickGraham101) [inline terraform/variables.tf:113]: this shouldn't have a default value, db name is required

File: terraform/app.tf (hunk 7 above)
Comment (gunndabad) [reply terraform/app.tf:114]: Production is using `SQL_Latin1_General_CP1_CI_AS` too

File: terraform/variables.tf (hunk 17 above)
Comment (gunndabad) [reply terraform/variables.tf:97]: >Is enabling private endpoint a requirement?

Don't think we need that

File: terraform/variables.tf (hunk 13 above)
Comment (enformatique11) [reply terraform/variables.tf:57]: I have added the description

File: terraform/app.tf (hunk 7 above)
Comment (enformatique11) [reply terraform/app.tf:113]: Do we need to change this or keep it

File: .github/workflows/build-and-deploy-notifications.yml (hunk 3 above)
Comment (enformatique11) [reply .github/workflows/build-and-deploy-notifications.yml:19]: removed

File: terraform/app.tf (hunk 7 above)
Comment (NickGraham101) [reply terraform/app.tf:113]: It can stay the same as test envs will match prod

File: .github/workflows/build-and-deploy-notifications.yml [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+
+      - name: Lint
+        run: |
+          #https://github.com/dotnet/format/issues/1433
+          dotnet tool install -g dotnet-format --version "7.*" --add-source https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet7/nuget/v3/index.json
+          dotnet-format --verify-no-changes
+        working-directory: DqtNotifications
+
+      - name: Build
+        run: dotnet build --configuration Release
```
Comment (NickGraham101) [inline .github/workflows/build-and-deploy-notifications.yml:54]: Missing `/p:NetCoreBuild=true` arg

File: terraform/variables.tf (hunk 15 above)
Comment (enformatique11) [reply terraform/variables.tf:78]: done

Comment (enformatique11) [reply terraform/variables.tf]: added prefix

File: terraform/variables.tf (hunk 19 above)
Comment (enformatique11) [reply terraform/variables.tf]: removed variable

**Review by gunndabad — APPROVED** [review]

**Review by NickGraham101 — APPROVED** [review]

**Pull request merged**